extension/js/common/core/crypto/smime/smime-key.ts

```diff
@@ -118,31 +197,67 @@ export class SmimeKey {
     } as Key;
-    (key as unknown as { rawArmored: string }).rawArmored = pem;
+    SmimeKey.saveArmored(key, certificateOrText, privateKey);
     return key;
   }
 
-  private static parsePemCertificate = (text: string): Key => {
-    const certificate = forge.pki.certificateFromPem(text);
-    return SmimeKey.getKeyFromCertificate(certificate, text);
+  private static getArmoredPrivateKey = (key: Key) => {
+    return (key as unknown as { privateKeyArmored: string }).privateKeyArmored;
+  }
+
+  private static getArmoredCertificate = (key: Key) => {
+    return (key as unknown as { certificateArmored: string }).certificateArmored;
+  }
+
+  private static getCertificate = (key: Key) => {
+    return forge.pki.certificateFromPem(SmimeKey.getArmoredCertificate(key));
   }
 
-  private static removeWeakKeys = (certificate: forge.pki.Certificate) => {
+  private static saveArmored = (key: Key, certificate: forge.pki.Certificate | string, privateKey: forge.pki.PrivateKey | string | undefined) => {
+    const armored = [];
+    if (privateKey) {
+      const armoredPrivateKey = (typeof privateKey === 'string') ? privateKey : forge.pki.privateKeyToPem(privateKey);
+      armored.push(armoredPrivateKey);
+      (key as unknown as { privateKeyArmored: string }).privateKeyArmored = armoredPrivateKey;
+    }
+    const armoredCertificate = (typeof certificate === 'string') ? certificate : forge.pki.certificateToPem(certificate);
+    armored.push(armoredCertificate);
+    (key as unknown as { certificateArmored: string }).certificateArmored = armoredCertificate;
+    (key as unknown as { rawArmored: string }).rawArmored = armored.join(''); // todo: crlf?
+  }
+
+  private static isKeyWeak = (certificate: forge.pki.Certificate) => {
     const publicKeyN = (certificate.publicKey as forge.pki.rsa.PublicKey)?.n;
     if (publicKeyN && publicKeyN.bitLength() < 2048) {
-      certificate.publicKey = undefined;
+      return true;
     }
-    const privateKeyN = (certificate.privateKey as forge.pki.rsa.PrivateKey)?.n;
-    if (privateKeyN && privateKeyN.bitLength() < 2048) {
-      certificate.privateKey = undefined;
+    return false;
+  }
+
+  private static checkPrivateKeyCertificateMatchOrThrow = (certificate: forge.pki.Certificate, privateKey: forge.pki.PrivateKey): void => {
+    const publicKeyN = (certificate.publicKey as forge.pki.rsa.PublicKey)?.n;
+    const publicKeyE = (certificate.publicKey as forge.pki.rsa.PublicKey)?.e;
+    const privateKeyN = (privateKey as forge.pki.rsa.PrivateKey).n;
+    const privateKeyE = (privateKey as forge.pki.rsa.PrivateKey).e;
+    if (publicKeyN && publicKeyE && privateKeyN && privateKeyE) {
+      // compare RSA Private and Public Key material
+      if (publicKeyN.compareTo(privateKeyN) === 0 && publicKeyE.compareTo(privateKeyE) === 0) {
+        return;
+      }
+      throw new Error("Certificate doesn't match the private key");
     }
+    throw new Error("This key type is not supported");
```

Comment:
Should we add support for ed25519 in this PR?

Reply:
Can be filed as a separate issue for later